File: Design a Stack With Increment Operation/Leetcode_1381.py
class CustomStack:

    def __init__(self, maxSize: int):
        self.stack = [None for _ in range(maxSize)]
        self.maxSize = maxSize
        self.n = 0

    def push(self, x: int) -> None:
        # print(f"pushing {x}, stack={self.stack}")
        if self.n < self.maxSize:
            self.stack[self.n] = x
            self.n += 1
        # print(f"stack following push: {self.stack}")

    def pop(self) -> int:
        if self.n > 0:
            value = self.stack[self.n-1]
            self.stack[self.n-1] = None
            self.n -= 1
        else:
            value = -1
        return value

    def increment(self, k: int, val: int) -> None:
        numElements = min(k, self.n)
        for i in range(numElements):
            self.stack[i] += val
        # print(self.stack)


# Your CustomStack object will be instantiated and called as such:
# obj = CustomStack(maxSize)
# obj.push(x)
# param_2 = obj.pop()
# obj.increment(k,val)

File: Design a Stack With Increment Operation/Leetcode_1381.cpp
class CustomStack {

private:
    int capacity;
    vector<int> stack;
    int n = 0;
    
public:
    CustomStack(int maxSize) {
        capacity = maxSize;
        vector<int> arr(capacity, 0);
        stack = arr;
    }
    
    void push(int x) {
        if (n < capacity){
            stack[n] = x;
            n++;
        }
    }
    
    int pop() {
        int value;
        if (n>0){
            value = stack[n-1];
            n--;}
        else {
            value=-1;
        }
        return value;
    }
    
    void increment(int k, int val) {
        for (int i=0; i<k and i<n; ++i){
            stack[i] += val;
        }
    }
};

/**
 * Your CustomStack object will be instantiated and called as such:
 * CustomStack* obj = new CustomStack(maxSize);
 * obj->push(x);
 * int param_2 = obj->pop();
 * obj->increment(k,val);
 */
